diff --git a/libHSAIL/libHSAIL/HSAILBrigObjectFile.cpp b/libHSAIL/libHSAIL/HSAILBrigObjectFile.cpp
index 905bb9d..b49b0fd 100644
--- a/libHSAIL/libHSAIL/HSAILBrigObjectFile.cpp
+++ b/libHSAIL/libHSAIL/HSAILBrigObjectFile.cpp
@@ -154,7 +154,7 @@ public:
             return 1;
         }
         if (fmt == FILE_FORMAT_AUTO) {
-            if (elfHeader.e_type == ET_REL && elfHeader.e_machine == EM_HSAIL_) {
+            if (elfHeader.e_type == ET_REL /*&& elfHeader.e_machine == EM_HSAIL_*/) {
                 fmt = FILE_FORMAT_BIF;
             } else if (elfHeader.e_type == ET_NONE && elfHeader.e_machine == EM_NONE) {
                 fmt = FILE_FORMAT_BRIG;
